partial/libuptiny/manifest.c

```diff
@@ -0,0 +1,114 @@
+#include "manifest.h"
+#include "base64.h"
+#include "crypto_api.h"
+#include "state_api.h"
+#include "utils.h"
+
+#include <string.h>
+
+static const char* attack_to_string(uptane_attack_t attack) {
+  switch (attack) {
+    case ATTACK_NONE:
+      return "";
+    case ATTACK_ROOT_THRESHOLD:
+      return "Failed threshold for root metadata";
+    case ATTACK_TARGETS_THRESHOLD:
+      return "Failed threshold for targets metadata";
+    case ATTACK_ROOT_VERSION:
+      return "Root rollback attempted";
+    case ATTACK_TARGETS_VERSION:
+      return "Targets rollback attempted";
+    case ATTACK_ROOT_EXPIRED:
+      return "Root metadata has expired";
+    case ATTACK_TARGETS_EXPIRED:
+      return "Targets metadata has expired";
+    case ATTACK_ROOT_LARGE:
+      return "Root metadata size exceeds the limit";
+    case ATTACK_TARGETS_LARGE:
+      return "Targets metadata size exceeds the limit";
+    case ATTACK_IMAGE_HASH:
+      return "Firmware image hash verification failed";
+    case ATTACK_IMAGE_LARGE:
+      return "Firmware image length mismatch";
+    default:
+      return "Unknown";
+  }
+}
+
+static const char* hash_alg_to_string(crypto_hash_algorithm_t alg) {
+  switch (alg) {
+    case CRYPTO_HASH_SHA512:
+      return "sha512";
+    default:
+      return "Unknown";
+  }
+}
+
+static const char* crypto_alg_to_method(crypto_algorithm_t alg) {
+  switch (alg) {
+    case CRYPTO_ALG_ED25519:
+      return "ed25519";
+    default:
+      return "Unknown";
+  }
+}
+
+/*{"attacks_detected":"<attack>","ecu_serial":"<serial>","installed_image":{"fileinfo":{"hashes":{"<hash_type>":"<hash
+ * in
+ * hex>"},"length":<length>},"filepath":"<target_name>"},"previous_timeserver_time":"1970-01-01T00:00:00Z","timeserver_time":"1970-01-01T00:00:00Z"}*/
+void uptane_write_manifest(char* signed_part, char* signatures_part) {
+  /* All strings are either literals or taken from trusted environment, use 'unsafe' functions */
```

Comment:
I'd recommend still taking the output buffer size into account. Then, with a test case using the maximum expected size in all fields, we'll get a good confidence that this will not overflow unintentionally.

Reply:
I don't really buy the optimization argument for this kind of check (without measurement). Also, as `strcpy` and `strcat` are strongly discouraged this day, it could end up begin the only function using them, which would in fact increase the size of the binary.

But ok, that's probably low risk here.